tempesta_fw/http_msg.c

```diff
@@ -549,50 +549,109 @@ tfw_http_msg_hdr_add(TfwHttpMsg *hm, TfwStr *hdr)
 	return __hdr_add(hm, hdr, hid);
 }
 
-void
-tfw_http_msg_iter_init(TfwHttpMsg *hm, TfwMsgIter *it)
+/**
+ * Allocate skb space for further @hm data writing.
+ * Put as much as possible to one skb, TCP GSO will care about segmentation.
+ *
+ * tfw_http_msg_free() is expected to be called for @hm if the function fails.
+ */
+static int
+__msg_alloc_skb_data(TfwHttpMsg *hm, size_t len)
 {
+	int i_skb, nr_skbs = DIV_ROUND_UP(len, SS_SKB_MAX_DATA_LEN);
+	struct sk_buff *skb;
+
+	for (i_skb = 0; i_skb < nr_skbs; ++i_skb) {
+		skb = ss_skb_alloc(min_t(size_t, len, SS_SKB_MAX_DATA_LEN));
+		if (!skb)
+			return -ENOMEM;
+		ss_skb_queue_tail(&hm->msg.skb_list, skb);
+	}
+
+	return 0;
+}
+
+/*
+ * Allocate an HTTP message of type @type and set it up with empty SKB
+ * space of size @data_len for data writing. An iterator @it is set up
+ * to support consecutive writes. This function is intended to work
+ * together with tfw_http_msg_write() that uses the @it iterator.
+ */
+TfwHttpMsg *
+tfw_http_msg_create(TfwMsgIter *it, int type, size_t data_len)
+{
+	TfwHttpMsg *hm = tfw_http_msg_alloc(type);
+
+	if ((hm == NULL) || (data_len == 0))
+		return NULL;
```

Comment:
It seems we should return either @hm if data_len == 0 (@hm is still valid new message, but with zero body) or free the message before return. I'd prefer the first one.